replica/table.cc

```diff
@@ -2937,13 +2937,31 @@ table::query(schema_ptr s,
         co_return make_lw_shared<query::result>();
     }
 
-    _async_gate.enter();
+    // 1 for the table and >= 1 for the compaction groups that overlap
+    utils::small_vector<gate::holder, 2> async_gate_holders;
+    async_gate_holders.push_back(_async_gate.hold());
+    {
+        // This looks bad but in the vast majority of the cases, there is a
+        // single range for a single partition in partition_ranges.
+        // This is the case we are optimizing for.
+        utils::small_vector<dht::token_range, 1> token_ranges;
+        for (const auto& pr : partition_ranges) {
+            token_ranges.push_back(pr.transform(std::mem_fn(&dht::ring_position::token)));
+        }
+        for (auto& cg : compaction_groups()) {
+            for (const auto& tr: token_ranges) {
```

Comment:
See below, this cartesian product can be huge.

Reply:
Fixed in v2.